# 01. Import Libraries

In [1]:
import pandas as pd
import numpy as np
import os

# 02. Import Data


In [2]:
path=r'/Users/nikolinastekovic/Library/CloudStorage/OneDrive-Personal/CF DATA ANALYICS/Immerse/4-Python/Instacart Basket Analysis'

In [3]:
path

'/Users/nikolinastekovic/Library/CloudStorage/OneDrive-Personal/CF DATA ANALYICS/Immerse/4-Python/Instacart Basket Analysis'

# 03. Task

### STEP 3. Import the orders_products_combined dataframe from the pickle file you just saved

In [10]:
df_ords_prod_comb = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_combined.csv'), index_col=False, encoding='ISO-8859-1')

Because of issues with importing .pkl format, I had to add this parametar "encoding='ISO-8859-1'" to the command read_csv.

### STEP 4. Check the shape of the imported dataframe

In [11]:
df_ords_prod_comb.shape

(32434489, 12)

Number of columns is not matching the merged table from the previous excercise. It should be (32434489, 11). I will run "head" command to take a look at the columns.

In [12]:
df_ords_prod_comb.head()

Unnamed: 0  order_id  user_id  order_number  Day_of_order  \
0           0   2539329        1             1             2   
1           1   2539329        1             1             2   
2           2   2539329        1             1             2   
3           3   2539329        1             1             2   
4           4   2539329        1             1             2   

   order_hour_of_day  days_since_prior_order  new_customer  product_id  \
0                  8                     NaN          True         196   
1                  8                     NaN          True       14084   
2                  8                     NaN          True       12427   
3                  8                     NaN          True       26088   
4                  8                     NaN          True       26405   

   add_to_cart_order  reordered _merge  
0                  1          0   both  
1                  2          0   both  
2                  3          0   both  
3                  4          0   both  
4                  5          0   both

In the process of export and import, Pandas created additional index colun. I will drop it, so the dataframe matched the original one.

In [13]:
df_ords_prod_comb = df_ords_prod_comb.drop(columns = 'Unnamed: 0', axis=1)

In [14]:
df_ords_prod_comb.head()

order_id  user_id  order_number  Day_of_order  order_hour_of_day  \
0   2539329        1             1             2                  8   
1   2539329        1             1             2                  8   
2   2539329        1             1             2                  8   
3   2539329        1             1             2                  8   
4   2539329        1             1             2                  8   

   days_since_prior_order  new_customer  product_id  add_to_cart_order  \
0                     NaN          True         196                  1   
1                     NaN          True       14084                  2   
2                     NaN          True       12427                  3   
3                     NaN          True       26088                  4   
4                     NaN          True       26405                  5   

   reordered _merge  
0          0   both  
1          0   both  
2          0   both  
3          0   both  
4          0   both

In [15]:
df_ords_prod_comb.shape

(32434489, 11)

### STEP 5. Determine a suitable way to combine the orders_products_combined dataframe with your products data set.

In [16]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'), index_col = False)

Products dataframe has to be imported and checked

In [17]:
df_prods.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

In [18]:
df_prods = df_prods.drop(columns = 'Unnamed: 0', axis=1)

In [19]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

Identifier column is product_id

Before merging I neeed to drop the column _merge from dataframe ords_prods_comb because when we merge a new dataframe, again a new _merge column will be created. That would interfere with the old one.


In [20]:
df_ords_prod_comb = df_ords_prod_comb.drop(columns = ['_merge'])

In [21]:
df_ords_prod_comb.shape

(32434489, 10)

In [23]:
df_merged = df_ords_prod_comb.merge(df_prods, on = 'product_id', indicator = True)

In [24]:
df_merged.head()

order_id  user_id  order_number  Day_of_order  order_hour_of_day  \
0   2539329        1             1             2                  8   
1   2398795        1             2             3                  7   
2    473747        1             3             3                 12   
3   2254736        1             4             4                  7   
4    431534        1             5             4                 15   

   days_since_prior_order  new_customer  product_id  add_to_cart_order  \
0                     NaN          True         196                  1   
1                    15.0         False         196                  1   
2                    21.0         False         196                  1   
3                    29.0         False         196                  1   
4                    28.0         False         196                  1   

   reordered product_name  aisle_id  department_id  prices _merge  
0          0         Soda        77              7     9.0   both  
1          1         Soda        77              7     9.0   both  
2          1         Soda        77              7     9.0   both  
3          1         Soda        77              7     9.0   both  
4          1         Soda        77              7     9.0   both

### STEP 6. Confirm the results of the merge using the merge flag.


In [25]:
df_merged['_merge'].value_counts()

_merge
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64

In [26]:
df_merged.shape

(32404859, 15)

###  STEP 7.  Export the newly created dataframe in a suitable format

Due to big size of the datafranme I will export it to .pkl fromat

In [27]:
df_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged.pkl'))